firmware/util/containers/type_list.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+template<typename base_t, typename... tail_t>
+struct type_list {
```

Comment:
smells like `std::variant` 🙂

Reply:
If variant is a union, type_list is a struct.  Items are non overlapping.

Reply:
yep - this is `tuple` but with a built-in version of `std::apply`

Reply:
well, kinda - tuple allows duplicates, and apply calls your callback once with all the values as parameters.  type_list::apply_all calls your callback once for each parameter, so you don't have to detangle the parameters yourself.  Although maybe ... would handle that for you?

Reply:
ok nevermind std::get does ensure type uniqueness if you get by type instead of index.  So I will remove a bunch of this and just use std::tuple; controllers can be wrapped in a Mockable<> which will have operator-> and a set method, and then have a custom apply method that makes the API nicer (unless I can make std::apply work well using ... somehow)

Reply:
arg, with some things Mockable<> and some not, it becomes a total mess.  you need to do get<mockable<foo>> and duplicate detection doesn't see through the mockable.

So, back to a custom implementation.  Maybe I can lean more on tuple/get though and make simpler wrappers.

Reply:
this is simple enough that I don't think we need std tuple